### PR TITLE
[MXNET-1227] Adding CornerPooling operator

### PR DESCRIPTION
**Pull request by BigDeviltjj**
## Description ##
Add operator:

1. CornerPooling operator

link to [JIRA issue 1227](https://issues.apache.org/jira/browse/MXNET-1227)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] CornerPooling
- [x] docs
- [x] unittest

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by stu1130** [comment]
@mxnet-label-bot add [pr-awaiting-review]
Thanks for doing this @BigDeviltjj

**Comment by BigDeviltjj** [comment]
It seems that the errors of checks was due to docker building, all builds on arm embedded platform were failed. However, I could not find any hint on the error of my code.

Could anyone tell me the reason of above problem please?

**Comment by Roshrini** [comment]
@apeforest @eric-haibin-lin Can you please review this PR?

File: tests/python/unittest/test_operator.py [hunk 1]
```diff
@@ -7231,6 +7231,18 @@ def f_sm_ce(data, label):
     np_one_hot_label[np.arange(batch_size), np_label] = 1.
     check_symbolic_forward(sym, {'data' : np_data, 'label' : np_label}, [np.array([f_sm_ce(np_sm, np_one_hot_label)])], rtol=1e-3, atol=1e-5)
 
+@with_seed()
+def test_corner_pooling():
+    data = mx.sym.Variable(name='data')
```
Comment (anirudhacharya) [inline tests/python/unittest/test_operator.py]: also call `check_symbolic_forward` and `check_symbolic_backward` - https://mxnet.incubator.apache.org/faq/add_op_in_backend.html#unit-test

Comment (BigDeviltjj) [reply tests/python/unittest/test_operator.py]: No problem, thank you!

Comment (BigDeviltjj) [reply tests/python/unittest/test_operator.py]: I have added these two functions, ```check_symbolic_forward``` and  ```check_symbolic_backward ```.

File: src/operator/contrib/corner_pooling-inl.h [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+inline void corner_pool(mshadow::Stream<cpu> *s,
+                        const DType *in_data,
+                        const TShape &ishape,
+                        const int corner_pooling_type,
+                        OpReqType req_type,
+                        DType *out_data) {
+  CHECK_EQ(req_type, kWriteTo)
+    << "Only support req=kWriteTo in corner pooling operations";
+  using mshadow::red::limits::MinValue;
+// const TShape &oshape = ishape;
```
Comment (anirudh2290) [inline src/operator/contrib/corner_pooling-inl.h]: remove unused code

File: src/operator/contrib/corner_pooling-inl.h [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+      h_step = -1;
+      h_start = height - 1;
+      h_end = -1;
+    } else {
+      h_step = +1;
+      h_start = 0;
+      h_end = height;
+    }
+    const index_t data_offset = width * height;
+    for (index_t b{0}; b < ishape[0]; ++b)
```
Comment (anirudh2290) [inline src/operator/contrib/corner_pooling-inl.h]: can you make it consistent with other mxnet code: ```for  (index_t b = 0; b < ishape[0]; ++b)``` here and at other places.

File: src/operator/contrib/corner_pooling-inl.h [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+      for (index_t c{0}; c < ishape[1]; ++c) {
+        for (index_t w{0}; w < width; ++w) {
```
Comment (anirudh2290) [inline src/operator/contrib/corner_pooling-inl.h]: can you add TODOs for optimization with ```Kernel::Launch```

File: src/operator/contrib/corner_pooling.cc [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  CornerPoolingParam param;
+  param.Init(attrs->dict);
+  attrs->parsed = param;
+}
+
+
+static bool CornerPoolingType(const nnvm::NodeAttrs &attrs,
+                              std::vector<int> *in_attrs,
+                              std::vector<int> *out_attrs) {
+  out_attrs->at(0) = in_attrs->at(0);
```
Comment (anirudh2290) [inline src/operator/contrib/corner_pooling.cc]: use ```TYPE_ASSIGN_CHECK```

Comment (BigDeviltjj) [reply src/operator/contrib/corner_pooling.cc]: I have modified all the code you mentioned above. Please check my code, Thank you so much and Happy Christmas Eve!

**Comment by Roshrini** [comment]
@anirudh2290 @anirudhacharya Can you take a look if your comments are addressed? Thanks!

**Comment by anirudhacharya** [comment]
My comments are addressed.

@BigDeviltjj can you please rebase push.

**Comment by BigDeviltjj** [comment]
@anirudhacharya I have rebased my push.

**Comment by sandeep-krishnamurthy** [comment]
@anirudhacharya - Can you please take a look back at this PR?

**Comment by anirudhacharya** [comment]
@sandeep-krishnamurthy As mentioned above my comments have already been addressed. If @anirudh2290's comments are addressed, this PR can be merged.

**Comment by vandanavk** [comment]
@BigDeviltjj could you resolve the branch conflict?

@anirudh2290 for review/merge

**Comment by anirudhacharya** [comment]
@anirudh2290 I see that your comments are addressed, can you please take another look at this PR.

File: src/operator/contrib/corner_pooling.cu [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    const int gridSize = (count + mshadow::cuda::kMaxThreadsPerBlock - 1) /
+                         mshadow::cuda::kMaxThreadsPerBlock;
+    dim3 dimGrid(mshadow::cuda::kMaxGridDim,
+                 (gridSize + mshadow::cuda::kMaxGridDim - 1) /
+                  mshadow::cuda::kMaxGridDim);
+    dim3 dimBlock(mshadow::cuda::kMaxThreadsPerBlock);
+    mshadow::cuda::CheckLaunchParam(dimGrid,
+            dimBlock, "Corner Pooling Forward");
+    cudaStream_t stream = mshadow::Stream<gpu>::GetStream(s);
+    CornerPoolingForwardTBKernel<DType><<<  dimGrid, dimBlock, 0, stream>>>
```
Comment (anirudh2290) [inline src/operator/contrib/corner_pooling.cu]: can we add launch bound guards similar to https://github.com/apache/incubator-mxnet/pull/13188

**Comment by karan6181** [comment]
@BigDeviltjj Can you please address the @anirudh2290 comments?
Also, please resolve the branch conflicts. Thanks!

**Comment by abhinavs95** [comment]
@BigDeviltjj ping again! Could you please address the new comments and rebase?

@mxnet-label-bot update [pr-awaiting-response]

**Comment by piyushghai** [comment]
@BigDeviltjj Gentle ping...

**Comment by Roshrini** [comment]
@BigDeviltjj Can you address few comments and resolve conflicts so that we can move forward with this PR?

**Comment by roywei** [comment]
@BigDeviltjj gentle ping to resolve the conflicts. thanks!

**Comment by vandanavk** [comment]
@mxnet-label-bot add [Operator]

**Comment by karan6181** [comment]
@BigDeviltjj Could you please provide an update on this PR? It's being more than 3 months and the PR is idle. If you are unable to work on this PR, you may want to consider closing it, and resubmitting once you have the bandwidth to work on it. Thank You!

**Pull request closed without merging**